ODBC fix for TIME/DATE/TIMESTAMP to solve .NET System.Data.ODBC issue take 2

bucweat opened the pull request:
While using some amazing git-fu wizardry trying to sync my fork to main, I managed to close my original PR #9551. I cannot figure out how to recover, so I'm submitting this new PR which contains the same changes as original, and also the changes requested by @maiadegraaf right before the inadvertent closing. 

@maiadegraaf also mentioned in PR #9551: Also would it be possible to add a test in `tools/odbc/test/tests/col_attribute/`, following the same structure as `int_query` and `char_query`, so that we can make sure all the column attributes are correctly set for the timestamp, date, and time. This is a great idea, however the PR with the col_attribute tests showed up after I submitted the original PR so at the time I didn't have access to that code. I've since merged in those changes, but would still like to save new col_attribute tests for a separate PR if that is ok with everyone.

I'm including some of comments and reference links from original PR #9551 below so that they are included with the PR that hopefully gets merged.

.NET System.Data.ODBC uses SQLGetData and SQLColAttribute under the hood. SQLColAttribute was returning the wrong SQL_DESC_CONCISE_TYPE which caused SQLGetData to fail. I was able to replicate the issue using direct calls to SQLGetData and SQLColAttribute and wrote some tests based on this (time.cpp).

> @Tishj: I would just like to see some references to the spec to support these changes Can you point out on https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcolattribute-function?view=sql-server-ver16 and https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlgetdata-function?view=sql-server-ver16 where this behavior is described?
> Maybe describing the current behavior and the new behavior would also help, for posterity

Like a lot of the ODBC documentation there is plenty of it but finding the answer you're looking for is not that easy...so maybe describing the detective process I followed might be helpful for posterity as you say...

When I was working through the problem, I wrote a test program that mimicked the method calls (SQLColAttribute/SQLFetch/SQLGetData) that I was seeing in the ODBC trace logs when I ran my .Net program that was currently failing. I found that in my test program, I could also accomplish the same task using SQLBindCol/SQLFetch and that was working. 

What was different? I was explicitly tell SQLBindCol what data type I was querying, while with SQLColAttribute/SQLGetData I was relying on SQLColAttribute returning the `SQL_DESC_CONCISE_TYPE` that I then passed to SQLGetData. In .NET these two operate under the hood so I'm assuming they work in tandem. I could see the value returned by SQLColAttribute in the ODBC trace logs as `9` regardless of if the query column datatype was TIMESTAMP, DATE, or TIIME. And I could see `9` passed to SQLGetData which failed (if I remember correctly DuckDB returned unhandled data type.

With query `SELECT TIMESTAMP '1992-09-20 11:30:00.123456'` you pass SQLBindCol `SQL_TYPE_TIMESTAMP` for the TargetType to get a `TIMESTAMP_STRUCT` back.  `SQL_TYPE_TIMESTAMP` = 93 which comes from sql.h. What if SQLColAttribute returned a `93` when the queried column was a TIMESTAMP?

I put together an app that linked in the ODBC driver directly so that I could debug the driver, and set to figure out where the `9` was coming from. I managed (with some trial and error too) to figure out that the `9` came from the `SQL_DATA_TYPE` column of `ApiInfo::ODBC_SUPPORTED_SQL_TYPES`. For TIMESTAMP (and DATE and TIME too) it contained `SQL_DATETIME` which from sql.h has a value of `9`. I replaced `SQL_DATETIME` with `SQL_TYPE_TIMESTAMP`, ran my test program, and received a valid timestamp. I then tried my .Net test program, and received a valid timestamp there also. 

Wash, rinse, repeat for DATE `91` and TIME `92`.

The root problem here is that SQLColAttribute was not returning the correct `SQL_DESC_CONCISE_TYPE`. I am surprised that others who are using ODBC have not experienced this issue before...

From sqlext.h
```
#define SQL_C_TYPE_DATE                 SQL_TYPE_DATE
#define SQL_C_TYPE_TIME                 SQL_TYPE_TIME
#define SQL_C_TYPE_TIMESTAMP            SQL_TYPE_TIMESTAMP
```

From sql.h
```
#define SQL_TYPE_DATE       91
#define SQL_TYPE_TIME       92
#define SQL_TYPE_TIMESTAMP  93
```
 
`SQL_DESC_CONCISE_TYPE` is discussed on https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcolattribute-function?view=sql-server-ver16. It points to https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/data-type-identifiers-and-descriptors?view=sql-server-ver16 but to be honest I didn't use that. https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/c-data-types?view=sql-server-ver16 was helpful for the C structs like TIMESTAMP_STRUCT used to decode values returned by SQLBindCol and SQLGetData. 

Another good reference:

https://learn.microsoft.com/en-us/sql/odbc/reference/appendixes/datetime-data-types?view=sql-server-ver16

which contains the following

> In ODBC 3.x, the identifiers for date, time, and timestamp SQL data types have changed from SQL_DATE, SQL_TIME, and SQL_TIMESTAMP (with instances of #define in the header file of 9, 10, and 11) to SQL_TYPE_DATE, SQL_TYPE_TIME, and SQL_TYPE_TIMESTAMP (with instances of #define in the header file of 91, 92, and 93), respectively. The corresponding C type identifiers have changed from SQL_C_DATE, SQL_C_TIME, and SQL_C_TIMESTAMP to SQL_C_TYPE_DATE, SQL_C_TYPE_TIME, and SQL_C_TYPE_TIMESTAMP, respectively, and the instances of #define have changed accordingly.

This might explain why System.Data.ODBC would get the date part right if the type was DATE  or TIMESTAMP as DuckDB was returning SQL_DATETIME == 9 == SQL_DATE. I'm assuming that given DuckDB reports version as ODBC 3 that 9 should be invalid, but maybe .Net tries to be compatible with previous versions and so tries to work with the 9 by processing as date?!?!? 

EDIT: I now believe that it is the OS ODBC plumbing (the magic that connects your app to the DuckDB ODBC driver) that is mapping the 9 to a 91. I added some cout statements to DuckDB ODBC driver without this PR changes to output SQL_DESC_CONCISE_TYPE and it definitely returns a 9, but in the ODBC Data Source Administrator trace log a 91 is returned by SQLColAttribute.

Commit:
fixing TIME/DATE/TIMESTAMP in .Net which uses SQLGetData and SQLColAttribute.

Commit:
Merge branch 'its_all_about_time' of https://github.com/bucweat/duckdb-bucweat into its_all_about_time

bucweat committed:
Merge branch 'duckdb:main' into its_all_about_time

Commit:
replace numerical values with macro values makes comparison test more obvious.

bucweat committed:
Merge branch 'duckdb:main' into its_all_about_time

bucweat committed:
apeasing the format CI.

Commit:
the way things used to be...

Commit:
Merge branch 'its_all_about_time' of https://github.com/bucweat/duckdb-bucweat into its_all_about_time

bucweat committed:
Merge branch 'duckdb:main' into its_all_about_time

bucweat committed:
apease the format CI.

maiadegraaf reviewed (commented):
Noticed a couple more things, like I said in the other PR adding `colattribute` tests would also be much appreciated!

Commit:
switch from (SQLCHAR *) to ConvertToSQLCHAR()

Commit:
the real fix is in...

Commit:
adding missing test files.

bucweat committed:
apease the format CI

bucweat committed:
more formatting woes...

bucweat committed:
more formattng woes...

bucweat commented:
The original fix in this PR (changing the values of SQL_DATA_TYPE in ODBC_SUPPORTED_SQL_TYPES) was not actually a fix but it illustrated a problem that SQL_DESC_CONCISE_TYPE was always being set to SQL_DESC_TYPE. For most data types this is ok. However, for DATETTIME types SQL_DESC_CONCISE_TYPE needs to be set separately per this reference:

https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlcolattribute-function?view=sql-server-ver16

The problem is that the code always sets SQL_DESC_CONCISE_TYPE = SQL_DESC_TYPE:

https://github.com/duckdb/duckdb/blob/c92602b864a09717b56843ae0811d85914509a9a/tools/odbc/descriptor.cpp#L807

This PR sets `sql_desc_type` and `sql_desc_concise_type` explicitly in `SetSqlDescType()`. I commented out the call to `SetSqlDataType()` as it doesn't seem necessary...`sql_desc_datetime_interval_code` is set after the call to SetSqlDataType() (which doesn't include the IsIntervalType() and GetIntervalCode() checks), so some refactoring may still be necessary. Also I left some comments in the code that could be removed later during said refactoring. 

I also added 4 new SQLColAttribute test files. Thanks to @maiadegraaf for providing original int/char versions these are based on and pushing to add these here...it did make me dig deeper and understand what was really going on. 

Also added is a SQL macro value LUT that makes the error reporting in TestAllFields() a little easier to work with.

maiadegraaf commented on tools/odbc/parameter_descriptor.cpp:
```diff
@@ -3,6 +3,8 @@
 #include "duckdb/common/types/decimal.hpp"
 #include "handle_functions.hpp"
 
+#include <iostream>
+
```
Nit: Is this neccisary?

bucweat replied:
Sorry left over from adding cout statements to driver.

maiadegraaf commented on tools/odbc/statement.cpp:
```diff
@@ -13,6 +13,8 @@
 
 #include <regex>
 
+#include <iostream>
+
```
Nit: Again this header is not in use.

bucweat replied:
ditto.

maiadegraaf commented on tools/odbc/statement_functions.cpp:
```diff
@@ -818,6 +818,7 @@ SQLRETURN duckdb::ExecuteStmt(SQLHSTMT statement_handle) {
 	return duckdb::BatchExecuteStmt(hstmt);
 }
 
+// https://learn.microsoft.com/en-us/sql/odbc/reference/syntax/sqlbindparameter-function?view=sql-server-ver16
```
This isn't the `SQLBindParameter` function, could you add this link to that function instead?

bucweat replied:
It's not...it's a DuckDB specific method, but the method signature matches so I thought the link is helpful. If you look at the SetDiagnosticRecord() messages all three have `SQLBindParameter`. Did some renaming happen at some point? 

What's more important is the [TODO](https://github.com/duckdb/duckdb/blob/f444a73249829caf5184324a178b3c2d81bb8131/tools/odbc/statement_functions.cpp#L856) listed further down...

maiadegraaf commented on tools/odbc/test/CMakeLists.txt:
```diff
@@ -26,6 +26,10 @@ add_executable(
   tests/col_attribute/col_attribute.cpp
   tests/col_attribute/char_query.cpp
   tests/col_attribute/int_query.cpp
+  tests/col_attribute/DATE_query.cpp
+  tests/col_attribute/TIME_query.cpp
+  tests/col_attribute/TIMESTAMP_query.cpp
+  tests/col_attribute/INTERVAL_query.cpp
```
Nit: Can these filenames match the case of the other filenames.

bucweat replied:
They do reference SQL data types and SQL convention is to upper case SQL keywords like data types....

maiadegraaf commented on tools/odbc/test/tests/col_attribute/DATE_query.cpp:
I appreciate you adding the comments to each attribute, could you find a centralized place (a readme or a larger comment somewhere that adds the information), instead of having it in every test, and only leaving the comments specific to the test?  I'm having a hard time reading the tests due to all the comments.

bucweat replied:
The comments are just snippets from the reference links...saves one from having to go to the link and find the info and explains why

maiadegraaf reviewed (commented):
Thanks for adding the `col_attribute` tests.  I'm happy it helped clear up where the bug was.

I left some initial comments and nits below. I'm curious to see what the tests do.

Furthermore,  we would prefer not to have any commented-out code in DuckDB so could you remove all of the code you commented out?

bucweat commented on tools/odbc/test/tests/col_attribute/utils.cpp:
```diff
@@ -11,7 +11,14 @@ void odbc_col_attribute_test::CheckString(SQLHANDLE handle, const std::string &e
 void odbc_col_attribute_test::CheckInteger(SQLHANDLE handle, SQLLEN expected, SQLSMALLINT field_identifier) {
 	SQLLEN number;
-	EXECUTE_AND_CHECK("SQLColAttribute", SQLColAttribute, handle, 1, field_identifier, nullptr, 0, nullptr, &number);
+	std::string state;
+	std::string message;
+	// EXECUTE_AND_CHECK("SQLColAttribute", SQLColAttribute, handle, 1, field_identifier, nullptr, 0, nullptr, &number);
+	SQLRETURN ret = SQLColAttribute(handle, 1, field_identifier, nullptr, 0, nullptr, &number);
+	if (ret != SQL_SUCCESS) {
+		ACCESS_DIAGNOSTIC(state, message, handle, SQL_HANDLE_STMT);
+	}
+	INFO(message);
```
IMHO the tests are not just for CI...they should be helpful to the developer too. In some cases where I needed to troubleshoot what was going on I had to unravel the fancy wrappings around the function calls as even with the debugger I could not tease out what was going on. Given that the CI is already pretty verbose as is, a little bit more can't hurt...much....

 As for where I put the INFO, if it was inside the if(){} then it is out of scope when the CHECK/REQUIRE gets called and so the INFO would not show up. So this was what I came up with for a compromise.

bucweat commented on tools/odbc/descriptor.cpp:
```diff
@@ -818,15 +823,22 @@ SQLRETURN DescRecord::SetSqlDataType(SQLSMALLINT type) {
 	auto type_info = vec_typeinfo.front();
+
 	// for consistency check set all other fields according to the first returned TypeInfo
-	SetSqlDataType(type_info.sql_data_type);
```
It is commented out in the PR code changes. 

This is why I added the INTERVAL test file...AFAICT everything is set correctly...but I'm just a journeyman so the more eyes on it the better...

maiadegraaf commented:
@Mytherin No, I'm still waiting for changes to be implemented.

bucweat commented:
> @Mytherin No, I'm still waiting for changes to be implemented.

Sorry but between popup work tasking and packing/moving back to USA where we will be on the road until 28 Dec, I don't have the cycles right now. Best guess for when I can reengage is sometime after 28 Dec when we move into our temporary housing.

Mytherin commented:
Closing this as it is stale
